### PR TITLE
[[ LCB Compiler ]] Generalized 'metadata' clause.

### PR DESCRIPTION
**Pull request by runrevmark**
The 'metadata' clause has been promoted to the level of a definition, meaning it can appear anywhere in a module after the 'use' clauses'.

The syntax of the metadata clause is now as follows:

```
   metadata NameOrString is String
```

The compiler does nothing with metadata clauses apart from pass them through to the manifest XML file.

Metadata clauses now appear as follows in the XML file:

```
 <metadata key="KEY">VALUE</metadata>
```

Note that a warning will be generated if metadata clauses are present before use clauses. Existing code should be updated to stop the warning.

### CODE REVIEW

**Comment by peter-b** [comment]
Why does `metadata` have to come after `use`?

**Comment by runrevmark** [comment]
Because as it stands 'use' is a separate syntactic thing than 'metadata'. Originally metadata was its own thing, but now its a definition. Previously you had to do:

```
    module ...
      { metadata }
      { use }
      { definitions }
   end module
```

Now the syntax is

```
   module ...
     { special case metadata }
     { use }
     { definitions }
  end module
```

Where definitions can contain metadata.

Existing code will continue to work with the special case, but the reason I put the warning in is because I am undecided as to whether to generalize 'use' to a definition level thing as well - which means it could be present anywhere in the module.

Basically hedging bets at the moment :)

**Comment by peter-b** [comment]
I think it's actually fair enough for `use` statements to be required to be before any definitions in the module.  There's a similar thing that catches people out in Scheme, whereby `define` can only used in "definition context".  I.e. your explanation makes sense, and I'm convinced that requiring `metadata` to appear after all `use` statements is reasonable. :+1:

**Comment by livecodestephen** [comment]
I'd prefer it if the meta data could go before the `use` statements.  I like a layout where a description of what something does comes before any implementation details; I say the module dependencies were an implementation detail so would come after the meta data.

**Comment by runrevmark** [comment]
Well it still can - for now. The metadata clause was added as an interim measure until we have some sort of 'higher-level' description of modules - allowing multiple source files per module, references to resources and code and such.

However, the 'metadata' clause could be set to become a general 'attribute' system for definitions I think - a bit like you get in C# and Java. So for that purpose it makes sense that they should be alongside the definitions they apply to.

(The reason for the generalization is so that you can specify sideline data next to properties for the IDE - like what editor to use in the property inspector).

I'm on the fence about how it might evolve at the moment - hence why you can still put metadata before use.

**Comment by peter-b** [comment]
Reviewed-By: Peter TB Brett peter@livecode.com
